README.rst
==============
pyramid_ipauth
==============

An authentication policy for Pyramid that sets identity and effective
principals based on the remote IP address of the request.


Overview
========

To perform IP-address-based authentication, create an IPAuthenticationPolicy
and specify the target IP range, userid and effective principals.  Then set it
as the authentication policy in your configurator::

    authn_policy = IPAuthenticationPolicy("127.0.*.*", "myuser", ["locals"])
    config.set_authentication_policy(authn_policy)

This will cause all requests from IP addresses in the 127.0.*.* range to be
authenticated as user "myuser" and have the effective principal "locals".

It is also possible to specify the configuration options in your deployment
file::

    [app:pyramidapp]
    use = egg:mypyramidapp

    ipauth.ipaddrs = 127.0.0.* 127.0.1.*
    ipauth.principals = locals

You can then simply include the pyramid_ipauth package into your configurator::

    config.include("pyramid_ipauth")

It will detect the ipauth settings and construct an appropriate policy.

Note that this package only supports matching against a single set of IP
addresss.  If you need to assign different credentials to different sets
of IP addresses, you can use the pyramid_multiauth package in conjunction
with pyramid_ipauth:

    http://github.com/mozilla-services/pyramid_multiauth

If you don't want to hard-code the userid or principals at configuration time,
you may specify a "get_userid" and/or "get_principals" callback instead.


Specifying IP Addresses
=======================

IP addresses can be specified in a variety of forms, including:

    * "all":        all possible IPv4 and IPv6 addresses
    * "local":      all local addresses of the machine
    * "A.B.C.D"     a single IP address
    * "A.B.C.D/N"   a network address specification
    * "A.B.C.*"     a glob matching against all possible numbers
    * "A.B.C.D-E"   a glob matching against a range of numbers
    * a whitespace- or comma-separated string of any of the above
    * a netaddr IPAddress, IPRange, IPGlob, IPNetork of IPSet object
    * a list, tuple or iterable of any of the above


Proxies
=======

This module does not respect the X-Forwarded-For header by default, since it
can be spoofed easily by malicious clients.  If your server is behind a 
trusted proxy that sets the X-Forwarded-For header, you should explicitly
declare the set of trusted proxies like so::

    IPAuthenticationPolicy("127.0.*.*",
                           principals=["local"],
                           proxies = "127.0.0.1")

The set of trusted proxy addresses can be specified using the same syntax as
the set of IP addresses to authenticate.


CHANGES.txt
0.3.3 - 2017-02-02
==================

- Fix splitting ipset strings on whitespace; thanks @kaleposhobios

0.3.2 - 2017-01-22
==================

- Resolve some warnings and clean up some tests; thanks Scott Searcy

0.3.1 - 2016-03-18
==================

- Fixes for python3 compatibility

0.3.0 - 2016-03-18
==================

- Add support for python3


0.2.0 - 2013-10-14
==================

- Add get_userid and get_principals callback functions; thanks mrijken
- Convert principals into a list if necessary; thanks janakj


0.1.1 - 2012-01-30
==================

- Update license to MPL 2.0


0.1.0 - 2011-11-11
==================

- Initial release


pyramid_ipauth/tests.py
# This Source Code Form is subject to the terms of the Mozilla Public
# License, v. 2.0. If a copy of the MPL was not distributed with this file,
# You can obtain one at http://mozilla.org/MPL/2.0/.

import unittest2

from netaddr import IPAddress, IPSet, IPRange, IPGlob, IPNetwork

import pyramid.testing
from pyramid.testing import DummyRequest
from pyramid.security import Everyone, Authenticated
from pyramid.interfaces import IAuthenticationPolicy

from pyramid_ipauth import IPAuthenticationPolicy, includeme
from pyramid_ipauth.utils import make_ip_set, get_ip_address, check_ip_address


class IPAuthPolicyTests(unittest2.TestCase):

    def setUp(self):
        self.config = pyramid.testing.setUp()

    def tearDown(self):
        pyramid.testing.tearDown()

    def test_remember_forget(self):
        policy = IPAuthenticationPolicy(["123.123.0.0/16"], "user")
        request = DummyRequest(environ={"REMOTE_ADDR": "192.168.0.1"})
        self.assertEqual(policy.remember(request, "user"), [])
        self.assertEqual(policy.forget(request), [])

    def test_remote_addr(self):
        policy = IPAuthenticationPolicy(["123.123.0.0/16"], "user")
        # Addresses outside the range don't authenticate
        request = DummyRequest(environ={"REMOTE_ADDR": "192.168.0.1"})
        self.assertEqual(policy.authenticated_userid(request), None)
        # Addresses inside the range do authenticate
        request = DummyRequest(environ={"REMOTE_ADDR": "123.123.0.1"})
        self.assertEqual(policy.authenticated_userid(request), "user")
        request = DummyRequest(environ={"REMOTE_ADDR": "123.123.1.2"})
        self.assertEqual(policy.authenticated_userid(request), "user")

    def test_noncontiguous_ranges(self):
        policy = IPAuthenticationPolicy(["123.123.0.0/16", "124.124.1.0/24"],
                                        "user")
        # Addresses outside the range don't authenticate
        request = DummyRequest(environ={"REMOTE_ADDR": "192.168.0.1"})
        self.assertEqual(policy.authenticated_userid(request), None)
        request = DummyRequest(environ={"REMOTE_ADDR": "124.124.0.1"})
        self.assertEqual(policy.authenticated_userid(request), None)
        # Addresses inside the range do authenticate
        request = DummyRequest(environ={"REMOTE_ADDR": "123.123.0.1"})
        self.assertEqual(policy.authenticated_userid(request), "user")
        request = DummyRequest(environ={"REMOTE_ADDR": "124.124.1.2"})
        self.assertEqual(policy.authenticated_userid(request), "user")

    def test_get_ip_address(self):
        # Testing without X-Forwarded-For
        request = DummyRequest(environ={"REMOTE_ADDR": "192.168.0.1"})
        self.assertEqual(get_ip_address(request),
                          IPAddress("192.168.0.1"))
        # Testing with X-Forwaded-For and no trusted proxies
        request = DummyRequest(environ={"REMOTE_ADDR": "192.168.0.1",
                                        "HTTP_X_FORWARDED_FOR": "123.123.0.1"})
        self.assertEqual(get_ip_address(request),
                          IPAddress("192.168.0.1"))
        # Testing with an untrusted proxy
        self.assertEqual(get_ip_address(request, "192.168.1.1"),
                          IPAddress("192.168.0.1"))
        # Testing with a trusted proxy
        self.assertEqual(get_ip_address(request, "192.168.0.1"),
                          IPAddress("123.123.0.1"))
        # Testing with a malformed XFF header
        request = DummyRequest(
            environ={
                "REMOTE_ADDR": "192.168.0.1",
                "HTTP_X_FORWARDED_FOR": "124.124.0.1 123.123.0.1"})
        self.assertEqual(get_ip_address(request, "192.168.0.1"),
                          IPAddress("192.168.0.1"))
        # Testing with a trusted proxy and untrusted proxy
        request = DummyRequest(
            environ={
                "REMOTE_ADDR": "192.168.0.1",
                "HTTP_X_FORWARDED_FOR": "124.124.0.1, 123.123.0.1"})
        self.assertEqual(get_ip_address(request, "192.168.0.1"),
                          IPAddress("123.123.0.1"))
        # Testing with several trusted proxies
        request = DummyRequest(
            environ={
                "REMOTE_ADDR": "192.168.0.1",
                "HTTP_X_FORWARDED_FOR": "124.124.0.1, 123.123.0.1"})
        self.assertEqual(get_ip_address(request, "192.168.0.1 123.123.0.1"),
                          IPAddress("124.124.0.1"))

    def test_check_ip_address(self):
        # Testing without X-Forwarded-For
        request = DummyRequest(environ={"REMOTE_ADDR": "192.168.0.1"})
        self.assertTrue(check_ip_address(request, "192.168.0.1"))
        self.assertTrue(check_ip_address(request, "192.168.0.1/8"))
        self.assertFalse(check_ip_address(request, "192.168.0.2"))

    def test_x_forwarded_for(self):
        policy = IPAuthenticationPolicy(["123.123.0.0/16"], "user",
                                        proxies=["124.124.0.0/24"])
        # Requests without X-Forwarded-For work as normal
        request = DummyRequest(environ={"REMOTE_ADDR": "192.168.0.1"})
        self.assertEqual(policy.authenticated_userid(request), None)
        request = DummyRequest(environ={"REMOTE_ADDR": "123.123.0.1"})
        self.assertEqual(policy.authenticated_userid(request), "user")
        # Requests with untrusted X-Forwarded-For don't authenticate
        request = DummyRequest(environ={"REMOTE_ADDR": "192.168.0.1",
                                        "HTTP_X_FORWARDED_FOR": "123.123.0.1"})
        self.assertEqual(policy.authenticated_userid(request), None)
        # Requests from single trusted proxy do authenticate
        request = DummyRequest(environ={"REMOTE_ADDR": "124.124.0.1",
                                        "HTTP_X_FORWARDED_FOR": "123.123.0.1"})
        self.assertEqual(policy.authenticated_userid(request), "user")
        # Requests from chain of trusted proxies do authenticate
        request = DummyRequest(
            environ={
                "REMOTE_ADDR": "124.124.0.2",
                "HTTP_X_FORWARDED_FOR": "123.123.0.1, 124.124.0.1"})
        self.assertEqual(policy.authenticated_userid(request), "user")
        # Requests with untrusted proxy in chain don't authenticate
        request = DummyRequest(
            environ={
                "REMOTE_ADDR": "124.124.0.1",
                "HTTP_X_FORWARDED_FOR": "123.123.0.1, 192.168.0.1"})
        self.assertEqual(policy.authenticated_userid(request), None)

    def test_principals(self):
        policy = IPAuthenticationPolicy(["123.123.0.0/16"],
                                        principals="test")
        # Addresses outside the range don't get metadata set
        request = DummyRequest(environ={"REMOTE_ADDR": "192.168.0.1"})
        self.assertEqual(policy.effective_principals(request), [Everyone])
        # Addresses inside the range do get metadata set
        request = DummyRequest(environ={"REMOTE_ADDR": "123.123.0.1"})
        self.assertEqual(policy.effective_principals(request),
                          [Everyone, "test"])
        policy.userid = "user"
        self.assertEqual(policy.effective_principals(request),
                          ["user", Everyone, Authenticated, "test"])

    def test_make_ip_set(self):
        def is_in(ipaddr, ipset):
            ipset = make_ip_set(ipset)
            return IPAddress(ipaddr) in ipset
        #  Test individual IPs
        self.assertTrue(is_in("127.0.0.1", "127.0.0.1"))
        self.assertFalse(is_in("127.0.0.2", "127.0.0.1"))
        #  Test globbing
        self.assertTrue(is_in("127.0.0.1", "127.0.0.*"))
        self.assertTrue(is_in("127.0.1.2", "127.0.*.*"))
        self.assertTrue(is_in("127.0.0.1", "127.0.0.*"))
        self.assertFalse(is_in("127.0.1.2", "127.0.0.*"))
        self.assertTrue(is_in("127.0.0.1", "127.0.0.1-5"))
        self.assertTrue(is_in("127.0.0.5", "127.0.0.1-5"))
        self.assertFalse(is_in("127.0.0.6", "127.0.0.1-5"))
        #  Test networks
        self.assertTrue(is_in("127.0.0.1", "127.0.0.0/8"))
        self.assertTrue(is_in("127.0.0.1", "127.0.0.0/16"))
        self.assertTrue(is_in("127.0.0.1", "127.0.0.0/24"))
        self.assertFalse(is_in("127.0.1.2", "127.0.0.0/24"))
        #  Test literal None
        self.assertFalse(is_in("127.0.0.1", None))
        #  Test special strings
        self.assertTrue(is_in("127.0.0.1", "local"))
        self.assertTrue(is_in("127.0.0.1", "all"))
        GOOGLE_DOT_COM = "74.125.237.20"
        self.assertFalse(is_in(GOOGLE_DOT_COM, "local"))
        self.assertTrue(is_in(GOOGLE_DOT_COM, "all"))
        #  Test with a list of stuff
        ips = ["127.0.0.1", "127.0.1.*"]
        self.assertTrue(is_in("127.0.0.1", ips))
        self.assertTrue(is_in("127.0.1.1", ips))
        self.assertFalse(is_in("127.0.0.2", ips))
        self.assertTrue(is_in("127.0.1.2", ips))
        #  Test with a string-list of stuff
        ips = "123.123.0.0/16 local"
        self.assertTrue(is_in("127.0.0.1", ips))
        self.assertTrue(is_in("127.0.1.1", ips))
        self.assertTrue(is_in("123.123.1.1", ips))
        self.assertFalse(is_in("124.0.0.1", ips))
        #  Test with list-splitting edge-cases
        self.assertTrue(is_in("127.0.0.1", "127.0.0.2,127.0.0.1"))
        self.assertTrue(is_in("127.0.0.1", "127.0.0.2,  127.0.0.1"))
        self.assertTrue(is_in("127.0.0.1", "127.0.0.2 127.0.0.1"))
        #  Test with various strange inputs to the parser
        self.assertTrue(is_in("127.0.0.1", IPAddress("127.0.0.1")))
        self.assertTrue(is_in("127.0.0.1", int(IPAddress("127.0.0.1"))))
        self.assertTrue(is_in("127.0.0.1", IPNetwork("127.0.0.1/8")))
        self.assertTrue(is_in("127.0.0.1", IPGlob("127.0.0.*")))
        self.assertTrue(is_in("127.0.0.1", IPRange("127.0.0.1", "127.0.0.2")))
        self.assertTrue(is_in("127.0.0.1", IPSet(["127.0.0.1/8"])))
        self.assertFalse(is_in("127.0.0.1", ""))
        self.assertFalse(is_in("127.0.0.1", None))
        self.assertRaises(ValueError, is_in, "127.0.0.1", 3.14159)
        self.assertRaises(ValueError, is_in, "127.0.0.1", Ellipsis)

    def test_callbacks(self):
        def get_userid(ipaddr):
            if str(ipaddr).startswith('192'):
                return 'LAN-user'
            if str(ipaddr).startswith('127'):
                return 'localhost-user'
            return None

        def get_principals(userid, ipaddr):
            principals = {
                'LAN-user': ['view'],
                'localhost-user': ['view', 'edit'],
                }
            return principals.get(userid, [])

        policy = IPAuthenticationPolicy("all", get_userid=get_userid,
                                        get_principals=get_principals)
        # Addresses outside the range don't authenticate
        request = DummyRequest(environ={"REMOTE_ADDR": "192.168.0.1"})
        self.assertEqual(policy.unauthenticated_userid(request), "LAN-user")
        self.assertEqual(policy.authenticated_userid(request), "LAN-user")
        self.assertEqual(policy.effective_principals(request),
                         ["LAN-user", Everyone, Authenticated, 'view'])
        request = DummyRequest(environ={"REMOTE_ADDR": "127.0.0.1"})
        self.assertEqual(policy.unauthenticated_userid(request),
                         "localhost-user")
        self.assertEqual(policy.authenticated_userid(request),
                         "localhost-user")
        self.assertEqual(policy.effective_principals(request),
                         ["localhost-user", Everyone, Authenticated,
                          'view', 'edit'])
        request = DummyRequest(environ={"REMOTE_ADDR": "86.8.8.8"})
        self.assertEqual(policy.unauthenticated_userid(request), None)
        self.assertEqual(policy.authenticated_userid(request), None)
        self.assertEqual(policy.effective_principals(request), [Everyone])

    def test_from_settings(self):
        settings = {
            "foo": "bar",
            "ipauth.ipaddrs": "123.123.0.1 124.124.0.1/16",
            "ipauth.userid": "one",
            "ipauth.principals": "two three",
            "otherauth.ipaddrs": "127.0.0.*",
            "otherauth.userid": "other",
            "otherauth.proxies": "127.0.0.1 127.0.0.2",
        }
        # Try basic instantiation.
        auth = IPAuthenticationPolicy.from_settings(settings)
        self.assertTrue(IPAddress("123.123.0.1") in auth.ipaddrs)
        self.assertEqual(auth.userid, "one")
        self.assertEqual(auth.principals, ["two", "three"])
        self.assertEqual(auth.proxies, IPSet([]))
        # Try instantiation with custom prefix.
        auth = IPAuthenticationPolicy.from_settings(settings,
                                                    prefix="otherauth.")
        self.assertTrue(IPAddress("127.0.0.1") in auth.ipaddrs)
        self.assertEqual(auth.userid, "other")
        self.assertEqual(auth.principals, [])
        self.assertTrue(IPAddress("127.0.0.1") in auth.proxies)
        self.assertTrue(IPAddress("127.0.0.2") in auth.proxies)
        self.assertFalse(IPAddress("127.0.0.3") in auth.proxies)
        # Try instantiation with extra keywords
        auth = IPAuthenticationPolicy.from_settings(settings,
                                                    prefix="otherauth.",
                                                    userid="overwritten")
        self.assertTrue(IPAddress("127.0.0.1") in auth.ipaddrs)
        self.assertEqual(auth.userid, "overwritten")
        self.assertEqual(auth.principals, [])
        self.assertTrue(IPAddress("127.0.0.1") in auth.proxies)
        self.assertTrue(IPAddress("127.0.0.2") in auth.proxies)
        self.assertFalse(IPAddress("127.0.0.3") in auth.proxies)

    def test_includeme(self):
        self.config.add_settings({"ipauth.userid": "user"})
        includeme(self.config)
        policy = self.config.registry.getUtility(IAuthenticationPolicy)
        self.assertEqual(policy.userid, "user")


pyramid_ipauth/utils.py
# This Source Code Form is subject to the terms of the Mozilla Public
# License, v. 2.0. If a copy of the MPL was not distributed with this file,
# You can obtain one at http://mozilla.org/MPL/2.0/.
"""

Utility functions for pyramid_ipauth

"""

import re
import socket

from netaddr import IPAddress, IPNetwork, IPGlob, IPRange, IPSet

from pyramid.compat import integer_types, string_types

#  This is used to split a string on an optional comma,
#  followed by any amount of whitespace.
_COMMA_OR_WHITESPACE = re.compile(r"(?:,\s*)|(?:\s+)")


def get_ip_address(request, proxies=None):
    """Get the originating IP address from the given request.

    This function resolves and returns the originating IP address of the
    given request, by looking up the REMOTE_ADDR and HTTP_X_FORWARDED_FOR
    entries from the request environment.

    By default this function does not make use of the X-Forwarded-For header.
    To use it you must specify a set of trusted proxy IP addresses.  The
    X-Forwarded-For header will then be traversed back through trusted proxies,
    stopping either at the first untrusted proxy or at the claimed original IP.
    """
    if proxies is None:
        proxies = IPSet()
    elif not isinstance(proxies, IPSet):
        proxies = make_ip_set(proxies)
    # Get the chain of proxied IP addresses, most recent proxy last.
    addr_chain = []
    try:
        xff = request.environ["HTTP_X_FORWARDED_FOR"]
    except KeyError:
        pass
    else:
        addr_chain.extend(a.strip() for a in xff.split(","))
    addr_chain.append(request.environ["REMOTE_ADDR"])
    # Pop trusted proxies from the list until we get the original addr,
    # or until we hit a malformed or untrusted proxy.
    addr = IPAddress(addr_chain.pop())
    while addr_chain:
        # If it's not a trusted proxy, stop the chain.
        if addr not in proxies:
            break
        # If next is a malformed IP address, stop the chain.
        if len(addr_chain[-1].split()) > 1:
            break
        addr = IPAddress(addr_chain.pop())
    return addr


def check_ip_address(request, ipaddrs, proxies=None):
    """Check whether a request originated within the given ip address set.

    This function checks whether the originating IP address of the request
    is within the given set of IP addresses, returning True if it is and False
    if not.

    By default this function does not make use of the X-Forwarded-For header.
    To use it you must specify a set of trusted proxy IP addresses which will
    be passed on to the get_ip_address function.
    """
    if not isinstance(ipaddrs, IPSet):
        ipaddrs = make_ip_set(ipaddrs)
    ipaddr = get_ip_address(request, proxies)
    return (ipaddr in ipaddrs)


def make_ip_set(ipaddrs):
    """Parse a variety of IP specifications into an IPSet object.

    This is a convenience function that allows you to specify a set of
    IP addresses in a variety of ways:

        * as an IPSet, IPAddress, IPNetwork, IPGlob or IPRange object
        * as the literal None for the empty set
        * as an int parsable by IPAddress
        * as a string parsable by parse_ip_set
        * as an iterable of IP specifications

    """
    # If it's already an IPSet, well, that's easy.
    if isinstance(ipaddrs, IPSet):
        return ipaddrs
    # None represents the empty set.
    if ipaddrs is None:
        return IPSet()
    # Integers represent a single address.
    if isinstance(ipaddrs, integer_types):
        return IPSet((IPAddress(ipaddrs),))
    # Strings get parsed as per parse_ip_set
    if isinstance(ipaddrs, string_types):
        return parse_ip_set(ipaddrs)
    # Other netaddr types can be converted into a set.
    if isinstance(ipaddrs, (IPAddress, IPNetwork)):
        return IPSet((ipaddrs,))
    if isinstance(ipaddrs, (IPGlob, IPRange)):
        return IPSet(ipaddrs.cidrs())
    # Anything iterable can be mapped over and unioned.
    try:
        ipspecs = iter(ipaddrs)
    except Exception:
        pass
    else:
        ipset = IPSet()
        for ipspec in ipspecs:
            ipset |= make_ip_set(ipspec)
        return ipset
    # Anything else is an error
    raise ValueError("can't convert to IPSet: %r" % (ipaddrs,))


def parse_ip_set(ipaddrs):
    """Parse a string specification into an IPSet.

    This function takes a string representing a set of IP addresses and
    parses it into an IPSet object.  Acceptable formats for the string
    include:

        * "all":        all possible IPv4 and IPv6 addresses
        * "local":      all local addresses of the machine
        * "A.B.C.D"     a single IP address
        * "A.B.C.D/N"   a network address specification
        * "A.B.C.*"     a glob matching against all possible numbers
        * "A.B.C.D-E"   a glob matching against a range of numbers
        * a whitespace- or comma-separated string of the above

    """
    ipset = IPSet()
    ipaddrs = ipaddrs.lower().strip()
    if not ipaddrs:
        return ipset
    for ipspec in _COMMA_OR_WHITESPACE.split(ipaddrs):
        # The string "local" maps to all local addresses on the machine.
        if ipspec == "local":
            ipset.add(IPNetwork("127.0.0.0/8"))
            for addr in get_local_ip_addresses():
                ipset.add(addr)
        # The string "all" maps to app IPv4 and IPv6 addresses.
        elif ipspec == "all":
            ipset.add(IPNetwork("0.0.0.0/0"))
            ipset.add(IPNetwork("::"))
        # Strings containing a "/" are assumed to be network specs
        elif "/" in ipspec:
            ipset.add(IPNetwork(ipspec))
        # Strings containing a "*" or "-" are assumed to be glob patterns
        elif "*" in ipspec or "-" in ipspec:
            for cidr in IPGlob(ipspec).cidrs():
                ipset.add(cidr)
        # Anything else must be a single address
        else:
            ipset.add(IPAddress(ipspec))
    return ipset


def get_local_ip_addresses():
    """Iterator yielding all local IP addresses on the machine."""
    # XXX: how can we enumerate all interfaces on the machine?
    # I don't really want to shell out to `ifconfig`
    # Sadly this is not guaranteed to succeed.
    try:
        for addr in socket.gethostbyname_ex(socket.gethostname())[2]:
            yield IPAddress(addr)
    except socket.gaierror:
        pass


pyramid_ipauth/__init__.py
# This Source Code Form is subject to the terms of the Mozilla Public
# License, v. 2.0. If a copy of the MPL was not distributed with this file,
# You can obtain one at http://mozilla.org/MPL/2.0/.
"""
IP-based authentication policy for pyramid.
"""
from zope.interface import implementer

from pyramid.interfaces import IAuthenticationPolicy
from pyramid.security import Everyone, Authenticated
from pyramid.authorization import ACLAuthorizationPolicy
from pyramid.settings import aslist
from pyramid.path import DottedNameResolver
from pyramid.compat import iteritems_, string_types

from pyramid_ipauth.utils import make_ip_set, check_ip_address, get_ip_address


__ver_major__ = 0
__ver_minor__ = 3
__ver_patch__ = 3
__ver_sub__ = ""
__ver_tuple__ = (__ver_major__, __ver_minor__, __ver_patch__, __ver_sub__)
__version__ = "%d.%d.%d%s" % __ver_tuple__


@implementer(IAuthenticationPolicy)
class IPAuthenticationPolicy(object):
    """An IP-based authentication policy for pyramid.

    This pyramid authentication policy assigns userid and/or effective
    principals based on the originating IP address of the request.

    You must specify a set of IP addresses against which to match, and may
    specify a userid and/or list of principals to apply.  For example, the
    following would authenticate all requests from the 192.168.0.* range as
    userid "myuser":

        IPAuthenticationPolicy(["192.168.0.0/24"], "myuser")

    The following would not authenticate as a particular userid, but would add
    "local" as an effective principal for the request (along with Everyone
    and Authenticated):

        IPAuthenticationPolicy(["127.0.0.0/24"], principals=["local"])

    By default this policy does not respect the X-Forwarded-For header since
    it can be easily spoofed.  If you want to respect X-Forwarded-For then you
    must specify a list of trusted proxies, and only forwarding declarations
    from these proxies will be respected:

        IPAuthenticationPolicy(["192.168.0.0/24"], "myuser",
                               proxies=["192.168.0.2"])

    Instead of given the IP addresses,
    userid and principals at configuration time.
    we can also compute these at runtime.

        def get_userid(ipaddr):
            "compute a userid based on the request"
            if str(ipaddr).startswith('192'):
                return 'LAN-user'
            return 'WAN-user'
        def get_principals(userid, ipaddr):
            "return a list of principals"
            if userid == 'WAN-user':
                return ['view']
            if userid == 'LAN-user':
                return ['view', 'edit']
            return []
        IPAuthenticationPolicy(get_userid=get_userid,
                                get_principals=get_principals)

    """

    def __init__(self, ipaddrs,
                 userid=None,
                 principals=None,
                 proxies=None,
                 get_userid=None,
                 get_principals=None):
        r = DottedNameResolver()
        self.get_userid = r.maybe_resolve(get_userid)
        self.get_principals = r.maybe_resolve(get_principals)
        self.ipaddrs = make_ip_set(ipaddrs)
        self.userid = userid
        if isinstance(principals, string_types):
            self.principals = aslist(principals)
        else:
            self.principals = principals
        self.proxies = make_ip_set(proxies)

    @classmethod
    def from_settings(cls, settings={}, prefix="ipauth.", **kwds):
        """Construct an IPAuthenticationPolicy from deployment settings."""
        # Grab out all the settings keys that start with our prefix.
        ipauth_settings = {}
        for name, value in iteritems_(settings):
            if not name.startswith(prefix):
                continue
            ipauth_settings[name[len(prefix):]] = value
        # Update with any additional keyword arguments.
        ipauth_settings.update(kwds)
        # Now look for specific keys of interest.
        ipaddrs = ipauth_settings.get("ipaddrs", "")
        userid = ipauth_settings.get("userid", None)
        principals = aslist(ipauth_settings.get("principals", ""))
        proxies = ipauth_settings.get("proxies", None)
        get_userid = ipauth_settings.get("get_userid", None)
        get_principals = ipauth_settings.get("get_principals", None)
        # The constructor uses make_ip_set to parse out strings,
        # so we're free to just pass them on in.
        return cls(ipaddrs, userid, principals,
                   proxies, get_userid, get_principals)

    def authenticated_userid(self, request):
        return self.unauthenticated_userid(request)

    def unauthenticated_userid(self, request):
        if not check_ip_address(request, self.ipaddrs, self.proxies):
            return None
        if self.get_userid is not None:
            userid = self.get_userid(get_ip_address(request, self.proxies))
        else:
            userid = self.userid
        return userid

    def effective_principals(self, request):
        principals = [Everyone]
        if not check_ip_address(request, self.ipaddrs, self.proxies):
            return principals

        if self.get_userid is not None:
            userid = self.get_userid(get_ip_address(request, self.proxies))
        else:
            userid = self.userid

        if userid is not None:
            principals.insert(0, userid)
            principals.append(Authenticated)

        if self.get_principals is not None:
            addr = get_ip_address(request, self.proxies)
            principals.extend(self.get_principals(userid, addr))
        elif self.principals is not None:
            principals.extend(self.principals)

        return principals

    def remember(self, request, principal, **kw):
        return []

    def forget(self, request):
        return []


def includeme(config):
    """Include default ipauth settings into a pyramid config.

    This function provides a hook for pyramid to include the default settings
    for ip-based auth.  Activate it like so:

        config.include("pyramid_ipauth")

    This will activate an IPAuthenticationPolicy instance with settings taken
    from the the application settings as follows:

        * ipauth.ipaddrs:     list of ip addresses to authenticate
        * ipauth.userid:      the userid as which to authenticate
        * ipauth.principals:  additional principals as which to authenticate
        * ipauth.proxies:     list of ip addresses to trust as proxies
        * ipauth.get_userid:    a (dotted name to a) callable which receive
                                the ip address to compute the userid
        * ipauth.get_principals:    a (dotted name to a) callable which receive
                                the ip address to compute the principals

    IP addresses can be specified in a variety of formats, including single
    addresses ("1.2.3.4"), networks ("1.2.3.0/16"), and globs ("1.2.3.*").
    You can also provide a whitespace-separated list of addresses to handle
    discontiguous ranges.
    """
    # Grab the pyramid-wide settings, to look for any auth config.
    settings = config.get_settings().copy()
    # Hook up a default AuthorizationPolicy.
    # ACLAuthorizationPolicy is usually what you want.
    # If the app configures one explicitly then this will get overridden.
    authz_policy = ACLAuthorizationPolicy()
    config.set_authorization_policy(authz_policy)
    # Use the settings to construct an AuthenticationPolicy.
    authn_policy = IPAuthenticationPolicy.from_settings(settings)
    config.set_authentication_policy(authn_policy)
